[feat-challenge-join] 챌린지 참가 기능 구현

71mC40H0 opened the pull request:
### 변경사항
<!-- 이 PR에서 어떤점들이 변경되었는지 기술해주세요. 가급적이면 as-is, to-be를 활용해서 작성해주세요.  -->
**AS-IS**
- 챌린지 참가 기능 구현
- 참기 인원이 다 차지 않고, 완료가 되지 않은 경우 참가 가능
- 회의 결과에 따라 생성한 Member의 id를 Challenge Entity에 추가

**TO-BE**
- 챌린지 나가기 기능 구현

### 테스트
<!-- 본 변경사항이 테스트가 되었는지 기술해주세요 --> 
- [x] 테스트 코드
- [x] API 테스트 

API 명세

#### 챌린지 참가
- challengeId: 1인 chllalenge 존재
- 해당 챌린지에 참가 중이지 않은 member로 요청

| 항목 | 요청 | 응답 | 비고 |
|----|----|----|----|
| 성공  |요청url: /challenge/join?challengeId=1  |{<br/>    "success": true,<br/>    "data": "Join Challenge Success"<br/>} | DB에 등록 확인  |
| 실패 - 인원이 다 찬 챌린지 | 요청url: /challenge/join?challengeId=1 | {<br/>    "success": false,<br/>    "data": "인원이 다 찼습니다."<br/>} | 인원이 다 찬 상태  |
| 실패 - 이미 종료된 챌린지 | 요청url: /challenge/join?challengeId=1 | {<br/>    "success": false,<br/>    "data": "이미 완료된 챌린지입니다."<br/>} | 챌린지 종료일이 지난 상태 |

71mC40H0 committed:
feat : 챌린지 참가 기능 구현

- 챌린지 인원이 다 차지 않고, 챌린지가 끝나지 않았다면 참가 가능
- 회의 결과에 따라 challenge에 생성한 memberId 추가

71mC40H0 committed:
test : 챌린지 참가 기능 테스트 코드 작성

71mC40H0 committed:
refactor : 응답내용 변경

- 완료 -> 종료로 수정

71mC40H0 committed:
test : 챌린지 참가 테스트 코드 추가

- 이미 종료된 챌린지의 경우 테스트 코드 추가

NavyHubb commented on src/main/java/com/zerototen/savegame/domain/entity/Challenge.java:
```diff
@@ -34,6 +34,8 @@ public class Challenge extends BaseEntity {
     @Column(name = "challenge_id")
     private Long id;
 
+    private Long createMemberId;
```
masterMemberId 가 나을듯합니다.

NavyHubb commented on src/main/java/com/zerototen/savegame/domain/entity/Challenge.java:
```diff
@@ -46,8 +48,9 @@ public class Challenge extends BaseEntity {
     private LocalDateTime deletedAt;
 
-    public static Challenge from(CreateChallengeServiceDto serviceDto) {
+    public static Challenge from(CreateChallengeServiceDto serviceDto, Long memberId) {
```
of

NavyHubb commented on src/main/java/com/zerototen/savegame/service/ChallengeService.java:
```diff
@@ -47,4 +49,41 @@ public ResponseDto<?> create(HttpServletRequest request, CreateChallengeServiceD
+
+        Member member = (Member) responseDto.getData();
+
+        Challenge challenge = challengeRepository.findById(challengeId).orElse(null);
```
줄바꿈

71mC40H0 committed:
refactor : 메소드명, url, 변수명 수정

- 매개변수가 2개 이상인 from 메소드의 이름을 of로 수정
- 챌린지 참가의 url을 PathVariable에서 RequestParam으로 수정
- Challenge Entity의 createMemberId를 masterMemeberId로 수정
- orElse문 한 줄 띄움

71mC40H0 committed:
test : 리팩토링에 따른 테스트 코드 수정
